lib/validates_overlap/overlap_validator.rb

```diff
@@ -90,19 +90,35 @@ def primary_key(record)
     record.class.primary_key
   end
 
+  def primary_key_value(primary_key_name, record)
+    record.send(primary_key_name)
+  end
+
+  def primary_key_type(record)
+    column_type(record.class,record.class.primary_key).to_sym
+  end
+
+  def column_type(model, name)
+   model.columns.detect { |c| c.name == name.to_s }.try(:type)
+  end
+
   # Generate sql condition for time range cross
   def generate_overlap_sql_conditions(record)
     starts_at_attr, ends_at_attr = attributes_to_sql(record)
     main_condition = condition_string(starts_at_attr, ends_at_attr)
     primary_key_name = primary_key(record)
+    key = primary_key_value(primary_key_name, record)
     if record.new_record?
       self.sql_conditions = main_condition
     else
-      self.sql_conditions = "#{main_condition} AND #{record_table_name(record)}.#{primary_key_name} != #{record.send(primary_key_name)}"
+      if primary_key_type(record) == :string
+        self.sql_conditions = "#{main_condition} AND #{record_table_name(record)}.#{primary_key(record)} != '#{key}'"
+      else
+        self.sql_conditions = "#{main_condition} AND #{record_table_name(record)}.#{primary_key(record)} != #{key}"
```

Comment:
Can you please refactor lines 115 and 117 to something like

```self.sql_conditions = "#{main_condition} AND #{record_table_name(record)}.#{primary_key(record)} !=```
```self.sql_conditions +=  primary_key_type(record) == :string ? '#{key}' : #{key}```

so we will remove the duplication.